osquery/remote/http/http_client.cpp

```diff
@@ -0,0 +1,317 @@
+void Client::sendRequest(STREAM_TYPE& stream,
+                         Request& req,
+                         beast_http_response_parser& resp) {
+  req.target((req.remotePath()) ? *req.remotePath() : "/");
+  req.version = 11;
+  req << Request::Header("Host", *client_options_.remote_hostname_);
+  req.prepare_payload();
+
+  req.keep_alive(true);
+  beast_http_request_serializer sr{req};
+  beast_http::write(stream, sr);
+
+  boost_asio::deadline_timer timer{ios_};
+  if (client_options_.timeout_) {
+    timer.expires_from_now(
+        boost::posix_time::seconds(client_options_.timeout_));
+    timer.async_wait(
+        [=](boost_system::error_code const& ec) { timeoutHandler(ec); });
+  }
+
+  boost::beast::flat_buffer b;
+  beast_http::async_read(
+      stream, b, resp, [&](boost_system::error_code const& ec) {
+        if (client_options_.timeout_) {
+          timer.cancel();
+        }
+        postResponseHandler(ec);
+      });
+
+  ios_.run();
+  ios_.reset();
+
+  if (ec_) {
+    throw std::system_error(ec_.value(), adapted_category(&ec_.category()));
+  }
+}
+
+Response Client::sendHTTPRequest(Request& req) {
+  do {
+    if (req.remoteHost()) {
+      client_options_.remote_hostname_ = *req.remoteHost();
+
+      if (req.remotePort()) {
+        client_options_.remote_port_ = *req.remotePort();
+      } else if (req.protocol() && (*req.protocol()).compare("https") == 0) {
+        client_options_.remote_port_ = std::to_string(HTTPS_DEFAULT_PORT);
+      } else {
+        client_options_.remote_port_ = std::to_string(HTTP_DEFAULT_PORT);
+      }
+
+      if (req.protocol()) {
+        if ((*req.protocol()).compare("https") == 0) {
+          ssl_connection = true;
+        } else {
+          ssl_connection = false;
+        }
+      } else {
+        ssl_connection = false;
+      }
+    }
+
+    beast_http_response_parser resp;
+    createConnection();
+
+    if (ssl_connection) {
+      encryptConnection();
+    }
+
+    if (ssl_connection) {
+      sendRequest(*ssl_sock_, req, resp);
+    } else {
+      sendRequest(sock_, req, resp);
+    }
+
+    switch (resp.get().result()) {
+    case beast_http::status::moved_permanently:
+    case beast_http::status::found:
+    case beast_http::status::see_other:
+    case beast_http::status::not_modified:
+    case beast_http::status::use_proxy:
+    case beast_http::status::temporary_redirect:
+    case beast_http::status::permanent_redirect: {
+      if (!client_options_.follow_redirects_) {
+        return Response(resp.release());
+      }
+
+      std::string redir_url = Response(resp.release()).headers()["Location"];
+      if (!redir_url.size()) {
+        throw std::runtime_error(
+            "Location header missing in redirect response.");
+      }
+
+      req.uri(redir_url);
+      LOG(INFO) << "HTTP(S) request re-directed to: " << redir_url;
+      break;
+    }
+    default:
+      return Response(resp.release());
+    }
+  } while (true);
```

Comment:
Is this request respecting a timeout? To prevent an infinite loop here?

Reply:
if client_options_.timeout_ is set then request times out see below -  
boost_asio::deadline_timer timer{ios_};
  if (client_options_.timeout_) {
    timer.expires_from_now(
        boost::posix_time::seconds(client_options_.timeout_));
    timer.async_wait(
        [=](boost_system::error_code const& ec) { timeoutHandler(ec); });
  }

Reply:
Good point. Yes, there is a chance of infinite loop here in case of if some malicious website continuously sends back the redirect in response. Earlier I had put individual aync_read under timeout. Now I have put the entire http request under timeout.

Reply:
I think you need to put a limit on the total number of redirections:
https://stackoverflow.com/questions/10895406/what-is-the-maximum-number-of-http-redirections-allowed-by-all-major-browsers

Reply:
After studying this issue, I went ahead with timeout solution.